### PR TITLE
[bugfix] _DimZeroAllToAll silently fails to send grads under torch.compile

### PR DESCRIPTION
**Pull request by stas00**
A Ulysses job whose decoder layers are `torch.compile`d delivers **no gradient at all** to the attention query, key and value projections. The forward pass is numerically correct, the loss is finite, no warning is raised, and attention simply stops learning: one optimizer step leaves those tensors bit-identical with their optimizer moments at exactly zero, while the output projection, the MLP and everything on the residual stream update normally.

The sequence/head exchange was a `torch.autograd.Function` whose forward allocated `torch.empty_like(input)` and filled it with `all_to_all_single`, a collective that writes into an output argument instead of returning a value. Dynamo does not treat an autograd function as an opaque call: it inlines the forward body and differentiates what it traced. The traced body allocates an empty tensor and mutates it through an operation the tracer does not model as producing data, so in the graph the exchange output carries no dependence on the exchange input, the derivative is exactly zero, and the hand-written backward -- the transposed exchange -- never runs. At execution time the collective still moves the bytes, which is why nothing about the forward looks wrong.

`_dim_zero_all_to_all` in `deepspeed/sequence/layer.py` replaces that function and calls `torch.distributed.nn.functional.all_to_all_single`, which returns the exchanged tensor. The dependence is then in the graph and the generated backward is the transposed exchange. Both call sites in `deepspeed/runtime/sequence_parallel/ulysses_sp.py` are updated; there is no other caller.

The commented-out line the previous implementation carried -- the same differentiable collective, called from inside the autograd function -- does not fix this. Holding everything else constant and varying only the implementation, on 2 and 4 ranks with the exchange inside `torch.compile`, against the eager gradient as reference:

| exchange implementation | compiled gradient |
| --- | --- |
| buffer-filling collective inside a `torch.autograd.Function` (previous) | zero |
| `torch.distributed.nn.functional.all_to_all_single` inside a `torch.autograd.Function` | zero |
| `torch.distributed.nn.functional.all_to_all_single` called directly (this PR) | matches eager |
| `torch.distributed._functional_collectives.all_to_all_single` inside a `torch.autograd.Function` | matches eager |
| `torch.distributed._functional_collectives.all_to_all_single` called directly | matches eager |

The two failing rows share the enclosing autograd function, which is what the tracer inlines. Removing it is the load-bearing part of the change; the public functional collective is preferred over `_functional_collectives` because it is not a private API.

Whether a given release inlines that body is a property of the tracer, not of this code: the same exchange on torch 2.9.1 keeps its gradient through a compiled region and loses it on torch 2.11.0. Code that reaches the wrong answer only on newer torch is worth fixing at the source rather than pinning around.

`single_all_to_all` in the same module has the same shape -- `torch.empty_like` followed by a mutating collective -- on the `DistributedAttention` path. It is not changed here: its `async_op=True` branch returns a work handle that a functional collective does not provide, so it needs a different treatment and its own measurement.

## Testing

`tests/unit/ulysses_alst/test_dim_zero_all_to_all.py` is new. Two ranks, both arms eager and compiled:

- The gradient test asserts an analytic value rather than a comparison against another run. The exchange is a permutation, so the derivative of `sum(exchange(x) ** 2)` is `2 * x` on every rank whichever slice a rank holds. The compiled arm of this test returns a zero gradient against the previous implementation and the exact value against this one.
- The forward test asserts that slot `i` of a rank's output holds the slot that rank owns in rank `i`'s input, so a substitution that is differentiable but exchanges the wrong data cannot pass.

The failing behaviour, measured on the previous implementation with a two-line driver -- one `torch.compile`d call of the exchange, loss `sum(y ** 2)`, H200, torch 2.11.0+cu130:

| ranks | eager gradient elements nonzero | compiled |
| --- | --- | --- |
| 2 | 64 / 64 | 0 / 64 |
| 4 | 128 / 128 | 0 / 128 |

The loss agrees to all printed digits between the two arms (`56.828213` on 2 ranks), which is the reason this is invisible in a training run.

Run against the previous exchange behind the new name, the test file reports `1 failed, 3 passed`: the compiled gradient arm fails with all 64 elements mismatched and a greatest absolute difference of 5.99, while the eager gradient arm and both forward arms pass. Against this change it reports `4 passed`. Two H200s, torch 2.11.0+cu130.

`tests/unit/ulysses_alst/test_ulysses_sp_hf.py` cannot be collected in the environment used here for a reason that predates this change: importing the module fails on a missing optional `transformers` dependency, before any DeepSpeed code runs.

End to end on a Hugging Face causal model with 24 attention heads, 4 key/value heads, head dimension 256, 1024 tokens, ZeRO-2, `sdpa`, decoder layers compiled: before the change `k_proj` and `v_proj` receive 0 of 5,242,880 gradient elements at sequence parallelism 2 and at 8; after it they receive 5,239,733 and 5,242,536. On a gated-attention model `q_proj` reads exactly half nonzero before the change, because the gate half of that projection is applied downstream of the exchange and keeps its gradient -- which is why a whole-tensor gradient norm on `q_proj` does not reveal the defect.

Also on this branch, unrelated to the above: `TiledMLP.backward` reshapes rather than views when flattening the activation and the incoming gradient, so a caller passing a non-contiguous tensor is not rejected, with a case in `tests/unit/ulysses_alst/test_tiled_compute.py`.

Not run: the `DistributedAttention` path described above, and any measurement of throughput change from the substituted collective.

### CODE REVIEW

File: deepspeed/sequence/layer.py [hunk 1]
```diff
@@ -10,6 +10,8 @@
 
 from einops import rearrange
 
+import torch.distributed.nn.functional as dist_nn
```
[the remark this answers is not in the document]
Comment (stas00) [reply deepspeed/sequence/layer.py:13]: the question is whether it'd still work, this is a very sensitive matter - the end result is that it has to come from torch.distributed.nn.functional

[the remark this answers is not in the document]
Comment (stas00) [reply deepspeed/sequence/layer.py:13]: @tjruwase, I have the patch to overcome this but it makes the code so much more complicated. thoughts?

[the remark this answers is not in the document]
Comment (stas00) [reply deepspeed/sequence/layer.py:13]:
```
diff --git a/deepspeed/runtime/sequence_parallel/ulysses_sp.py b/deepspeed/runtime/sequence_parallel/ulysses_sp.py
index e1468943d..8ea3895b6 100644
--- a/deepspeed/runtime/sequence_parallel/ulysses_sp.py
+++ b/deepspeed/runtime/sequence_parallel/ulysses_sp.py
@@ -31,7 +31,7 @@ https://github.com/snowflakedb/ArcticTraining/blob/main/projects/sequence-parall

 from collections import defaultdict, deque
 from deepspeed.runtime.utils import see_memory_usage
-from deepspeed.sequence.layer import _dim_zero_all_to_all
+from deepspeed.sequence.layer import _dim_zero_all_to_all, register_all_to_all_group
 from deepspeed.utils.logging import logger
 from einops import rearrange
 from packaging import version
@@ -105,6 +105,9 @@ class UlyssesSPAttentionHF(torch.nn.Module):
         super().__init__()
         self.attn = attn
         self.process_group = process_group
+        # the exchange operator resolves its group by name, and the registry has to be written here
+        # rather than at call time, which may be inside a compiled region
+        register_all_to_all_group(process_group)
         self.world_size = dist.get_world_size(process_group)
         self.sp_rank = dist.get_rank(process_group)

diff --git a/deepspeed/sequence/layer.py b/deepspeed/sequence/layer.py
index 52f12af55..8717adbf9 100644
--- a/deepspeed/sequence/layer.py
+++ b/deepspeed/sequence/layer.py
@@ -4,14 +4,12 @@
 # DeepSpeed Team
 import torch

-from typing import Any, Tuple
+from typing import Any, Dict, Tuple
 from torch import Tensor
 from torch.nn import Module

 from einops import rearrange

-import torch.distributed.nn.functional as dist_nn
-
 import deepspeed.comm as dist
 from deepspeed.accelerator import get_accelerator
 from deepspeed.module_inject.tp_shard import AutoTPMeta, get_shard_size_list
@@ -309,24 +307,64 @@ def single_all_to_all(input,
     return res


+# A process group cannot appear in an operator schema, so the exchange addresses its group by name and
+# resolves the name back to the group when it runs. The registry is written once, when the attention module
+# is built: a write issued from inside a compiled region does not land before the operator reads it.
+_all_to_all_groups: Dict[str, dist.ProcessGroup] = {}
+
+
+def register_all_to_all_group(group: dist.ProcessGroup) -> str:
+    """Make ``group`` reachable by name from inside the exchange operator, and return that name.
+
+    Call this outside any compiled region, before the first exchange on ``group``.
+    """
+    name = group.group_name
+    _all_to_all_groups[name] = group
+    return name
+
+
+@torch.library.custom_op("deepspeed::dim_zero_all_to_all", mutates_args=())
+def _dim_zero_all_to_all_op(input: Tensor, group_name: str) -> Tensor:
+    output = torch.empty_like(input)
+    dist.all_to_all_single(output, input, group=_all_to_all_groups[group_name])
+    return output
+
+
+@_dim_zero_all_to_all_op.register_fake
+def _(input: Tensor, group_name: str) -> Tensor:
+    return torch.empty_like(input)
+
+
+def _dim_zero_all_to_all_setup_context(ctx: Any, inputs: Tuple[Any, ...], output: Tensor) -> None:
+    ctx.group_name = inputs[1]
+
+
+def _dim_zero_all_to_all_backward(ctx: Any, grad_output: Tensor) -> Tuple[Tensor, None]:
+    return _dim_zero_all_to_all_op(grad_output.contiguous(), ctx.group_name), None
+
+
+_dim_zero_all_to_all_op.register_autograd(_dim_zero_all_to_all_backward,
+                                          setup_context=_dim_zero_all_to_all_setup_context)
+
+
 def _dim_zero_all_to_all(group: dist.ProcessGroup, input: Tensor) -> Tensor:
     """Differentiable All2All across dimension 0.

-    The collective has to return the exchanged tensor rather than fill a buffer the caller allocated. The


+# A process group cannot appear in an operator schema, so the exchange addresses its group by name and
+# resolves the name back to the group when it runs. The registry is written once, when the attention module
+# is built: a write issued from inside a compiled region does not land before the operator reads it.
+_all_to_all_groups: Dict[str, dist.ProcessGroup] = {}
+
+
+def register_all_to_all_group(group: dist.ProcessGroup) -> str:
+    """Make ``group`` reachable by name from inside the exchange operator, and return that name.
+
+    Call this outside any compiled region, before the first exchange on ``group``.
+    """
+    name = group.group_name
+    _all_to_all_groups[name] = group
+    return name
+
+
+@torch.library.custom_op("deepspeed::dim_zero_all_to_all", mutates_args=())
+def _dim_zero_all_to_all_op(input: Tensor, group_name: str) -> Tensor:
+    output = torch.empty_like(input)
+    dist.all_to_all_single(output, input, group=_all_to_all_groups[group_name])
+    return output
+
+
+@_dim_zero_all_to_all_op.register_fake
+def _(input: Tensor, group_name: str) -> Tensor:
+    return torch.empty_like(input)
+
+
+def _dim_zero_all_to_all_setup_context(ctx: Any, inputs: Tuple[Any, ...], output: Tensor) -> None:
+    ctx.group_name = inputs[1]
+
+
+def _dim_zero_all_to_all_backward(ctx: Any, grad_output: Tensor) -> Tuple[Tensor, None]:
+    return _dim_zero_all_to_all_op(grad_output.contiguous(), ctx.group_name), None
+
+
+_dim_zero_all_to_all_op.register_autograd(_dim_zero_all_to_all_backward,
+                                          setup_context=_dim_zero_all_to_all_setup_context)
+
+
 def _dim_zero_all_to_all(group: dist.ProcessGroup, input: Tensor) -> Tensor:
     """Differentiable All2All across dimension 0.

-    The collective has to return the exchanged tensor rather than fill a buffer the caller allocated. The
-    difference is invisible in eager mode and decisive under ``torch.compile``: a collective that writes into
-    an output argument leaves the traced graph holding an output with no data dependence on the input, so the
-    generated backward produces a zero gradient while the forward stays numerically correct, and training
-    silently stops updating everything upstream of the exchange.
+    The gradient of this exchange has to be something the tracer knows about. An autograd function that fills
+    a caller-allocated buffer with a collective is not: ``torch.compile`` inlines the forward body rather than
+    treating the call as opaque, and the traced body allocates an empty tensor and mutates it through an
+    operation that does not register as producing data, so the graph holds an output with no dependence on the
+    input. Differentiating that yields a zero gradient and the hand-written backward never runs, while the
+    forward stays numerically correct -- a training job silently stops updating everything upstream of the
+    exchange.

-    Wrapping the buffer-filling form in a ``torch.autograd.Function`` does not protect it. The tracer inlines
-    the function body instead of treating the call as opaque, and differentiates what it traced, so the
-    hand-written backward never runs.
+    Registering the exchange as an operator with a declared backward gives the tracer a derivative instead of
+    a body to inline, and keeps the collective on ``deepspeed.comm`` so the configured backend dispatches it.
     """
     world_size = dist.get_world_size(group)
     assert input.shape[0] == world_size, f"Dim 0 {input.shape[0]} is not world size"

-    input = input.contiguous()
-    return dist_nn.all_to_all_single(torch.empty_like(input), input, group=group)
+    return _dim_zero_all_to_all_op(input.contiguous(), group.group_name)


 class _SeqAllToAll(torch.autograd.Function):
diff --git a/tests/unit/ulysses_alst/test_dim_zero_all_to_all.py b/tests/unit/ulysses_alst/test_dim_zero_all_to_all.py
index 0956ca443..f035c33c4 100644
--- a/tests/unit/ulysses_alst/test_dim_zero_all_to_all.py
+++ b/tests/unit/ulysses_alst/test_dim_zero_all_to_all.py
@@ -7,7 +7,7 @@ UlyssesPlus: the dimension-zero all-to-all that trades sequence shards for head
 """

 from deepspeed.accelerator import get_accelerator
-from deepspeed.sequence.layer import _dim_zero_all_to_all
+from deepspeed.sequence.layer import _dim_zero_all_to_all, register_all_to_all_group
 from unit.common import DistributedTest
 from unit.util import torch_assert_close, torch_assert_equal
 import deepspeed.comm as dist
@@ -26,6 +26,7 @@ class TestDimZeroAllToAll(DistributedTest):
         the call is inside a compiled region, because the traced graph has no data dependence to differentiate.
         """
         group = dist.new_group(ranks=list(range(self.world_size)))
+        register_all_to_all_group(group)
         device = get_accelerator().current_device_name()

         torch.manual_seed(1234)
@@ -43,6 +44,7 @@ class TestDimZeroAllToAll(DistributedTest):
     def test_forward_moves_the_expected_slices(self, compile_exchange):
         """Slot ``i`` of this rank's output holds the slot this rank owns in rank ``i``'s input."""
         group = dist.new_group(ranks=list(range(self.world_size)))
+        register_all_to_all_group(group)
         device = get_accelerator().current_device_name()
         rank = dist.get_rank(group)

```

[the remark this answers is not in the document]
Comment (stas00) [reply deepspeed/sequence/layer.py:13]: there is a further complication - torch.library.custom_op needs torch 2.4 which would make this code even more complicated since it needs a fallback. And ds requires torch >2.0 - time to raise this lower bound to a higher one - keep say 2 years worth of torch versions supported?

[the remark this answers is not in the document]
Comment (sfc-gh-truwase) [reply deepspeed/sequence/layer.py:13]: @stas00 we decided a while back to tighten the lower bound to -2 minor versions. For example,  >=2.12.* currently given 2.14.0 is latest torch version. 

@loadams did we document this?

[the remark this answers is not in the document]
Comment (sfc-gh-truwase) [reply deepspeed/sequence/layer.py:13]: > [@tjruwase](https://github.com/tjruwase), I have the patch to overcome this but it makes the code so much more complicated. thoughts?

@stas00 i think the complicated code is worthwhile to maintain accelerator agnostic behavior.

File: deepspeed/sequence/layer.py [hunk 2]
```diff
@@ -307,24 +309,24 @@ def single_all_to_all(input,
...
 
 
-class _DimZeroAllToAll(torch.autograd.Function):
-    """Differentiable All2All across dimension 0."""
-
-    @staticmethod
-    def forward(ctx: Any, group: dist.ProcessGroup, input: Tensor) -> Tensor:
-        world_size = dist.get_world_size(group)
-        assert input.shape[0] == world_size, f"Dim 0 {input.shape[0]} is not world size"
+def _dim_zero_all_to_all(group: dist.ProcessGroup, input: Tensor) -> Tensor:
```
Comment (sfc-gh-truwase) [inline deepspeed/sequence/layer.py:312]: @neeldani, @tohtana I am curious if we encountered this silent grad issue in AutoSP?

File: deepspeed/sequence/layer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (stas00) [reply deepspeed/sequence/layer.py:13]: Really? -2 versions is a way too tight, I don't agree it's a good idea. 

2 years at least is my recommendation, which usually means 6 versions.

But in any case, it says 2.0.0 here:
https://github.com/deepspeedai/DeepSpeed/blob/71d316d608a56af2fcc27b84b14cf854b7052eff/requirements/requirements.txt#L10

which forces integrations to respect the requirement and backport to old pytorch.

File: deepspeed/sequence/layer.py (hunk 2 above)
Comment (stas00) [reply deepspeed/sequence/layer.py:312]: In general I'd recommend dropping all the original collective autograd functions in Deepspeed which were written when pytorch didn't have them and replace with torch.functional version of each, it'd be simpler and better for future proofing.

I already proposed doing that when porting Ulysses to the newer version, a year ago, but at the end decided to leave things as they are being unaware those functions silently failed torch.compile and destroyed grads.

File: deepspeed/sequence/layer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (stas00) [reply deepspeed/sequence/layer.py:13]: I'm away till Wed, where I can push the complex code, but please feel free to add the patch I shared above before that. We would need to make a new release after the merge.

[the remark this answers is not in the document]
Comment (sfc-gh-truwase) [reply deepspeed/sequence/layer.py:13]: > 2 years at least is my recommendation, which usually means 6 versions.

@stas00 , sounds great, -6 it is

[the remark this answers is not in the document]
Comment (stas00) [reply deepspeed/sequence/layer.py]: @sfc-gh-truwase, I pushed the abstraction changes as requested.

**Review by sfc-gh-truwase — APPROVED** [review]
